### PR TITLE
Attached media field

### PR DESCRIPTION
**Pull request by matiasmolleja**
This adds a new editor to the media field.

When you use this editor the media items:
- They are uploaded in place from the content item editor. Not picked from the assets browser.
- They are stored on an special folder dedicated to hold all attached media field's assets.
- That folder is only visible to users that have the "ManageAttachedMediaFieldsFolder" permission.
- Each content type will have its own subdir inside this special folder. I use an "splitting" dir name mechanism to avoid having a very large number of dirs.

I would need some guidance about a few things:

**1. How to keep clean the file system:** 
I need to remove temporary upload files.
And I think we should remove the dirs for the content items that are removed.
How to do that? In my first attempt I was removing temporary files through a background task.
Is that a good approach? 

**2. The dir name splitting algorithm**
Maybe we should make it optional through a setting using the new configuration system?

I'm using a 2 chars segment size. Is that OK? Should we use a longer segment?
Right now if an admin users needs to browse for a content item dir, it is a pain.

### CODE REVIEW

**Comment by sebastienros** [comment]
Can you use a single temporary folder that is common to all tenants? Like in the actual TEMP folder of the system with a well-known subfolder name, then delete this folder when the application starts up, and ensure you clean the temporary file once the item is published? This way even if items are uploaded but not used, we clean on startups, and usually items that are uploaded are then saved are correctly cleaned up. This way we don't need a bg task. 

You can split the folder in bigger chunks I assume. We could also have a custom behavior to display the list by display a flat list, while knowing that it's stored in a different physical file.

**Comment by matiasmolleja** [comment]
Nice.
And what about deleting content items? Maybe we should delete the assets once the content item is deleted? 
If yes. Do we need to take into account the drafts?  I mean delete the the assets only when they are not used anywhere.

**Comment by sebastienros** [comment]
Ideally each version should have their own file. And deleting a version should delete the files attached to it. Then we could say that it will take too much space. A solution could be to use file names that are hashes of their content (like git) such that if the file is the same, then the same file is reused. And when no other version is pointing to a file, then delete it. At least unique per content item. This way, when a content item is deleted, we can look at all versions for a link to this file, if non exist then we can delete it. I assume we should be fine with a folder like /contentype/contentitemId and if it gets too big suggest users to migrate to a store that is not bound to the filesystem.

**Comment by matiasmolleja** [comment]
About keeping clean the temp folder. 
Not sure how to create a folder common to all tenants, But I can investigate it. I just need to learn more :)
I think we can have tenants that use Azure together with tenants that use the server fileStore? Maybe that would be an issue if we use that approach?

I'm thinking on another possible solution. Let me know what do you think: 
Right now when the user publish the media field, I'm taking the assets from the temp folder, and moving them to the definitive content item id folder .
What about adding a check at this point, enumerate files on the temp folder, and deleting temp files older than a given time?

**Comment by agriffard** [comment]
@matiasmolleja  The size 160 is important because it is in the DefaultSizes https://github.com/OrchardCMS/OrchardCore/blob/d918b13c9a269e2e7b6280a9c2a8d9e69a368cf7/src/OrchardCore.Modules/OrchardCore.Media/Startup.cs#L53

so please, do not change it back to 120 or the thumbnails won't be processed.

**Comment by sebastienros** [comment]
`Environment.GetTempFolder()` or something similar. Then create a known sub-directory in it.

**Comment by matiasmolleja** [comment]
When the user clicks publish, the files on the temp folder are moved to the definitive folder on the mediafilestore. 
If we use azure storage, the temp folder would still be on the asp.net server right? If the user uploads a medium or large file there will be a potentially long delay until the file is moved from the asp.net server to the azure storage?

**Comment by sebastienros** [comment]
Correct. So I assume you can keep using the media store for temporary files, only delete them when the content is saved, and at startup. No background task. Should be easy based on what is already done.

**Comment by Skrypt** [comment]
Hi @matiasmolleja I sent you a message on Gitter. I'm trying to use _contentItemDisplayManager.BuildEditorAsync(contentItem, this, false) on a frontend securized page. Though the Vue.js component doesn't seem to like the path of files since it doesn't find any files. I think there is something wrong with the path it is using since it works but only on the admin side.

**Comment by matiasmolleja** [comment]
@sebastienros  do you know how to fix the conflicts?

**Comment by Skrypt** [comment]
Conflict happens with the jest functional test.

**Comment by sebastienros** [comment]
Status?

**Comment by Skrypt** [comment]
Working fine for me since a month on my website. Nothing to say about it else than it rocks 👍 @matiasmolleja

**Comment by matiasmolleja** [comment]
I did not merged the latest changes, It was working well last time I tried. I will try tonight.

**Comment by matiasmolleja** [comment]
After trying this with the latest changes It works except for the tenants.

When we start a non default tenant, we call the constructor in MediaFileStore, passing a "mediaUrlBase".
After merging the latest changes on this PR we are passing always "media", instead of "tenant1/media".
I think that maybe it has to do with the latest changes applied here #3200 ?

https://github.com/OrchardCMS/OrchardCore/blob/3325a9abaa347c508f69134b3af3a16a054c8762/src/OrchardCore.Modules/OrchardCore.Media/Startup.cs#L89

Before that we calculated mediaUrlBase with this code:
`var mediaUrlBase = "/" + fileStore.Combine(shellSettings.RequestUrlPrefix, AssetsUrlPrefix);`

@jtkech  could you please take a look and give me some hints about how to fix it?

**Comment by jtkech** [comment]
Yes `ConfigureServices()` is called before building the tenant pipeline but we init `mediaUrlBase` by using the `PathBase` inside an implementation factory which is executed after when 1st resolving the `IMediaFileStore` service.

Just tried on my side with the last dev branch, it works.

Media

**Comment by jtkech** [comment]
Okay, i understand.

You inject `IMediaFileStore` in `TempDirCleanerService` which is an `IModularTenantEvents` which is resolved before building the pipeline when activating the shell. And you can do that.

This was to make it working under a virtual folder, maybe i saw this because first i was using a more complex way to extract the virtual folder regardless of a current prefix in the pathBase, but forgot it and found this simple way.

Anyway good catch, i will fix it tomorrow.

**Comment by Skrypt** [comment]
This will require a merge of the dev branch as there is a lot of changes made in the startup.cs file because of recent fixes on the ImageSharp issues.

**Comment by matiasmolleja** [comment]
So I should wait for the PR from @jtkech  and then merge dev branch , test everything works and update this PR.

**Comment by matiasmolleja** [comment]
I tried and it works nicely after the fix provided by @jtkech  (Thanks!!!)

File: src/OrchardCore.Modules/OrchardCore.Media/Views/MediaField-Attached.Edit.cshtml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+<partial name="Shared/MediaFieldEditResources.cshtml" />
+
+<script at="Foot">
+
+    initializeAttachedMediaField(
+        document.getElementById('@mediaFieldId'),
+        '@mediaFieldId-field-file-upload',
+        '@Url.Action("Upload", "Admin", new { area = "OrchardCore.Media" }) ',
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media/Views/MediaField-Attached.Edit.cshtml]: Don't inject `@` inside javascript. Create hidden divs that have `data` attributes with the value, and load it from javascript.

File: src/OrchardCore.Modules/OrchardCore.Media/Views/MediaField.Edit.cshtml [hunk 2]
```diff
@@ -2,27 +2,22 @@
...
     var settings = Model.PartFieldDefinition.Settings.ToObject<MediaFieldSettings>();
     var modalMediaId = "Modal" + Html.IdFor(m => m);
 }
 
+<partial name="Shared/MediaFieldEditResources.cshtml" />
+
 <script at="Foot">
     initializeMediaApplication(false, '@Url.Action("MediaApplication", "Admin", new { area = "OrchardCore.Media" })');
-    initializeMediaFieldEditor(
+    initializeMediaField(
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media/Views/MediaField.Edit.cshtml]: Here too and everywhere there is an `@` in javascript

File: src/OrchardCore.Modules/OrchardCore.Media/Views/Shared/MediaFieldEditLocalization.cshtml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+@*Localizable Strings *@
+<input type="hidden" id="onlyOneItemMessage" value="@T["Only one item is allowed on this field."]" />
+
+@*Localizable Strings for mediaFieldThumbsContainerComponent*@
+<input type="hidden" id="t-media-not-found" value="@T["Not Found"]" />
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media/Views/Shared/MediaFieldEditLocalization.cshtml]: Yes, like this, better than injecting values in javascript

**Comment by matiasmolleja** [comment]
@sebastienros  added changes after review

**Pull request merged**